packages/api/amplify_api/lib/src/api_plugin_impl.dart

```diff
@@ -0,0 +1,397 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+library amplify_api;
+
+import 'dart:io';
+
+import 'package:amplify_api/amplify_api.dart';
+import 'package:amplify_api/src/graphql/helpers/send_graphql_request.dart';
+import 'package:amplify_api/src/graphql/providers/app_sync_api_key_auth_provider.dart';
+import 'package:amplify_api/src/graphql/providers/oidc_function_api_auth_provider.dart';
+import 'package:amplify_api/src/graphql/ws/web_socket_connection.dart';
+import 'package:amplify_api/src/native_api_plugin.dart';
+import 'package:amplify_api/src/util/amplify_api_config.dart';
+import 'package:amplify_api/src/util/amplify_authorization_rest_client.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
+
+/// {@template amplify_api.amplify_api_dart}
+/// The AWS implementation of the Amplify API category.
+/// {@endtemplate}
+class AmplifyAPIDart extends AmplifyAPI {
+  /// {@macro amplify_api.amplify_api_dart}
+  AmplifyAPIDart({
+    List<APIAuthProvider> authProviders = const [],
+    AWSHttpClient? baseHttpClient,
+    this.modelProvider,
+    this.subscriptionOptions,
+  })  : _baseHttpClient = baseHttpClient,
+        super.protected() {
+    authProviders.forEach(registerAuthProvider);
+  }
+
+  late final AWSApiPluginConfig _apiConfig;
+  final AWSHttpClient? _baseHttpClient;
+  late final AmplifyAuthProviderRepository _authProviderRepo;
+  final _logger = AmplifyLogger.category(Category.api);
+
+  /// A map of the keys from the Amplify API config with auth modes to HTTP clients
+  /// to use for requests to that endpoint/auth mode. e.g. { "myEndpoint.AWS_IAM": AWSHttpClient}
+  final Map<String, AWSHttpClient> _clientPool = {};
+
+  /// A map of the keys from the Amplify API config websocket connections to use
+  /// for that endpoint.
+  final Map<String, WebSocketConnection> _webSocketConnectionPool = {};
+
+  /// The registered [APIAuthProvider] instances.
+  final Map<APIAuthorizationType, APIAuthProvider> _authProviders = {};
+
+  /// Subscription options
+  final GraphQLSubscriptionOptions? subscriptionOptions;
+
+  @override
+  Future<void> configure({
+    AmplifyConfig? config,
+    required AmplifyAuthProviderRepository authProviderRepo,
+  }) async {
+    final apiConfig = config?.api?.awsPlugin;
+    if (apiConfig == null) {
+      throw const ApiException(
+        'No AWS API config found',
+        recoverySuggestion: 'Add API from the Amplify CLI. See '
+            'https://docs.amplify.aws/lib/graphqlapi/getting-started/q/platform/flutter/#configure-api',
+      );
+    }
+    _apiConfig = apiConfig;
+    _authProviderRepo = authProviderRepo;
+    _registerApiPluginAuthProviders();
```

Comment:
These should be registered in `addPlugin` to avoid race conditions

Reply:
The issue I was having there is that API key depends on the config which is not available in `addPlugin` (unless we just wanted to always register an API key provider whether or not it's in the config bc we do not read the value of API key here). The race condition was more of an issue when providers are used in other plugins so I did not think it would be an issue here, although at the very least a comment explaining why might be nice.